Add Pipelines

dvf opened the pull request:
# Ability to have Pipelines

Add Pipelines, so that you can modify the behaviour after a successful redirect. This is better (IMHO) than assuming that a user needs to be logged in.

If `SLACK_PIPELINES` is defined in settings, then the following default behaviour is ignored:

- Assume user logged in
- Redirect to success page


# Usage

Define pipelines in `settings.py`:

```
...
SLACK_PIPELINES = [
    'path.to.callable1',
    'path.to.callable2',
]
```

where a callable looks like: 

```
def callable1(request, api_data):
    do_something()
    ...
    return request, api_date
```

dvf committed:
Ability to have pipelines

izdi commented:
Great idea, thanks for the support and description. 
Could you provide info about new feature in `README.md` ?

izdi commented on django_slack_oauth/views.py:
```diff
@@ -57,15 +59,22 @@ def get(self, request, *args, **kwargs):
         if not api_data['ok']:
             return self.error_message(api_data['error'])
 
+        pipelines = settings.SLACK_PIPELINES
```
It would be better to call this as a separate helper function or method like `apply_pipelines` or smth.

dvf replied:
yeah, i'll review it after xmas :)

dvf committed:
Updated Readme

dvf committed:
Update README.md

dvf committed:
Refined Pipeline Behaviour

- Removed LoginRequired and relegated default behaviour to a pipeline
which created a request.
- Added default success view
- Updated Documentation
- Added Admin view

dvf commented:
Ok, I made a bit of a large update:

1. Updated Documentation 
2. Created a Logo (what do you think?)

Then I made some big assumptions:

3. User doesn't have to be logged in.
4. Created new model `SlackOAuthRequest`, which just logs requests.
5. New users can be created using a pipeline (added some examples to Documentation). 

I suggest we start adding some tests too

dvf commented:
@izdi Do you want me to add some more examples?

izdi commented on django_slack_oauth/models.py:
```diff
@@ -5,21 +5,17 @@
-    'SlackUser',
-)
 
+class SlackOAuthRequest(models.Model):
```
What about backward compatibility ? 
Renaming just removes the old table. Not a great situation for lib users.

dvf replied:
I can leave the old model and just move the functionality to a pipeline method to keep backward functionality.

izdi replied:
Yep, that would be great.

izdi commented:
Hi, sorry for the delay.

In general I appreciate the effort and thank you. This is great PR. 💯 
However I got a few question regarding changes. Check the comments.

izdi commented on django_slack_oauth/views.py:
```diff
@@ -1,43 +1,43 @@
 import uuid
+from importlib import import_module
+
+from django.contrib.auth.mixins import LoginRequiredMixin
```
Please format imports according to style guide
https://docs.djangoproject.com/en/dev/internals/contributing/writing-code/coding-style/

izdi commented on django_slack_oauth/pipelines.py:
```diff
@@ -0,0 +1,14 @@
+def log_request(request, api_data):
+        SlackOAuthRequest.objects.create(
+            access_token=api_data.pop('access_token'),
+            extras=api_data
```
Also IP address from request?

dvf replied:
Yeah, I'll remove.

izdi replied:
Remove ? :)
Since you added IP field in the model you could just add this on creation.

dvf replied:
@izdi On second thought, getting the IP is pretty opinionated and depends on the deployment/setup. For example, do we get it from `HTTP_X_FORWARDED_FOR` or from `REMOTE_ADDR`? If we're hosting behind an ELB, then we can trust the `HTTP_X_FORWARDED_FOR` but only if we've set up NGINX correctly. With some bad intentions, a user could fake it. See: http://stackoverflow.com/questions/18264304/get-clients-real-ip-address-on-heroku#answer-18517550

izdi commented on django_slack_oauth/views.py:
```diff
@@ -1,43 +1,43 @@
 from . import settings
-from .models import SlackUser
-
-__all__ = (
```
Let's not leave this behind ;)

izdi commented on django_slack_oauth/admin.py:
```diff
@@ -0,0 +1,8 @@
+from django.contrib import admin
+
+from django_slack_oauth.models import SlackOAuthRequest
```
local django import

dvf replied:
ah. good catch.

izdi commented on django_slack_oauth/pipelines.py:
```diff
@@ -0,0 +1,14 @@
+from django_slack_oauth.models import SlackOAuthRequest
```
local django import

dvf commented:
@izdi I've gone through your comments. I'll make some changes

dvf committed:
Cleaned up imports

dvf committed:
Backward Compatibility

dvf commented:
We should add some tests, but LGTM

dvf committed:
Fixed typo

izdi commented on django_slack_oauth/models.py:
```diff
@@ -18,8 +15,23 @@ class SlackUser(models.Model):
         db_table = 'slack_user'
+
+
+class SlackOAuthRequest(models.Model):
```
So now we have two tables.
`SlackUser` is never used anymore which feels wrong to me. I'd rather added new fields in there, this way is more compatible and users bound to `slack_user` will not have to change their code.

Could I get your intentions on why creating new table with almost exact set of fields ?

dvf replied:
@izdi, I think we should deprecate `SlackUser` because depending on the scopes that you define, the request may not be creating a user at all. So it seems more general to call it a `SlackOAuthRequest`. 

Also, don't you think it's better to let Django namespace the table instead of forcing a name? This way, if the user already has a `slack_user` table—which may be likely since they are building a Slack app—we don't cause conflicts.

izdi replied:
Agree about the naming. Just afraid for users about sudden stop of writes into `slack_user` :) On the other hand there was no any major release so version could be 1.0.0 with release notes about updates.

dvf replied:
@izdi, agreed on the version update. It makes things a lot simpler and gives us a chance to implement.

izdi commented on django_slack_oauth/views.py:
```diff
@@ -1,43 +1,46 @@
+
+    def get(self, request):
+        messages.success(request, "You've been successfully authenticated.")
+        return HttpResponse("Slack OAuth login successful.")
```
So redirecting from `error_message` method in `SlackAuthView` will lead to wrong UX, i.e. the error occurred but the message is success.

https://github.com/izdi/django-slack-oauth/pull/17/files#diff-247cca211a4c6d0d8c626e68b068eea1R57

dvf replied:
It's wrong UX anyway to add an error and redirect to a success URL? Perhaps it would be better if we defined a `ERROR_REDIRECT_URL` too, in the same way?

izdi replied:
You are correct about naming, but in your case if error occurs won't we be redirected with message `"You've been successfully authenticated."` Correct me if i'm wrong.
On the other hand `ERROR_REDIRECT_URL` could also be the referer.

dvf replied:
@izdi, yes. The message will be set at the point that the response != 200 but you will also see the success message at the view. If we're going to include this as a major version update I think we should define `ERROR_REDIRECT_URL` and avoid adding messages altogether and delegate that to a pipeline: `def add_success_messages()...` etc.

dvf commented:
Any thoughts?

izdi reviewed (approved):
Hey, in general I approve the changes. You can merge the PR. However I'd like you to add error link instead of bug above. After that I'll push new version to PYPI and a tag with new feature and description. 

I am still not sure that it's good practice to leave SlackUser model behind without writing to it. As a workaround I'd like to see a pipeline same as `log_request` but for SlackUser (not added to pipelines by default) in case users willing to use only it.

Also I gave you access to writes, as I strongly feel you are more into it right now. :1st_place_medal: 

Happy new year btw!

dvf commented:
@izdi awesome. Happy New Year! Apologies for the delay, been very busy with work :)
